# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/AudioStream.cpp
# Commit: f604fe0f8d02
# Full Hash: f604fe0f8d0275c12a17fa492948871fdd9f59bf
# Author: Paul Adenot <paul@paul.cx>
# Date: 2022-05-10 03:50:31
# Regressor Bug: 1743834
# File Overlap Count: 1
# Description:
#   Bug 1743834 - Wait for the audio callbacks to start being called before using the audio clock. r=alwu,media-playback-reviewers
#   
#   Because we're generally using high latency on the cubeb stream used by
#   AudioStream instance, it can take some time for the callbacks to start being
#   called, and the for the audio clock (cubeb_stream_get_position(...)) to advance.
# ==============================================================================

diff -r 61ec6368ddcd -r f604fe0f8d02 dom/media/AudioStream.cpp
--- a/dom/media/AudioStream.cpp	Mon May 09 13:20:50 2022 +0000
+++ b/dom/media/AudioStream.cpp	Mon May 09 13:20:50 2022 +0000
@@ -148,7 +148,9 @@
       mSandboxed(CubebUtils::SandboxEnabled()),
       mPlaybackComplete(false),
       mPlaybackRate(1.0f),
-      mPreservesPitch(true) {}
+      mPreservesPitch(true),
+      mCallbacksStarted(false)
+    {}
 
 AudioStream::~AudioStream() {
   LOG("deleted, state %d", mState.load());
@@ -231,8 +233,8 @@
   auto startTime = TimeStamp::Now();
   TRACE("AudioStream::Init");
 
-  LOG("%s channels: %d, rate: %d", __FUNCTION__, mOutChannels,
-      mAudioClock.GetInputRate());
+  LOG("%s channels: %d, rate: %d", __FUNCTION__, mOutChannels, mAudioClock.GetInputRate());
+
   mSinkInfo = aSinkInfo;
 
   cubeb_stream_params params;
@@ -594,6 +596,9 @@
     CubebUtils::GetAudioThreadRegistry()->Register(mAudioThreadId);
   }
   WebCore::DenormalDisabler disabler;
+  if (!mCallbacksStarted) {
+    mCallbacksStarted = true;
+  }
 
   TRACE_AUDIO_CALLBACK_BUDGET(aFrames, mAudioClock.GetInputRate());
   TRACE("AudioStream::DataCallback");